statsmodels/regression/process_reg.py

```diff
@@ -0,0 +1,716 @@
+
+        model = sm.OLS(self.endog, self.exog)
+        result = model.fit()
+
+        par = np.concatenate(
+            (result.params,
+             np.zeros(self.exog_scale.shape[1] + self.exog_smooth.shape[1])))
+
+        return par
+
+    def loglike(self, params):
+        """
+        Calculate the log-likelihood function for the model.
+
+        Parameters
+        ----------
+        params : array-like
+            The packed parameters for the model.
+
+        Returns
+        -------
+        The log-likelihood value at the given parameter point.
+
+        Notes
+        -----
+        The mean, scaling, and smoothing parameters are packed into
+        a vector.  Use `unpack` to access the component vectors.
+        """
+
+        mnpar, scpar, smpar = self.unpack(params)
+
+        # Residuals
+        resid = self.endog - np.dot(self.exog, mnpar)
+
+        # Scaling parameters
+        sc = np.exp(np.dot(self.exog_scale, scpar))
+
+        # Smoothness parameters
+        sm = np.exp(np.dot(self.exog_smooth, smpar))
+
+        # Get the log-likelihood
+        ll = 0.
+        for _, ix in self._groups_ix.items():
+
+            # Get the covariance matrix for this person.
+            cm = self.cov.get_cov(self.time[ix], sc[ix], sm[ix])
+
+            re = resid[ix]
+            ll -= 0.5 * np.linalg.slogdet(cm)[1]
+            ll -= 0.5 * np.dot(re, np.linalg.solve(cm, re))
+
+        if self.verbose:
+            print("L=", ll)
+
+        return ll
+
+    def score(self, params):
+        """
+        Calculate the score function for the model.
+
+        Parameters
+        ----------
+        params : array-like
+            The packed parameters for the model.
+
+        Returns
+        -------
+        The score vector at the given parameter point.
+
+        Notes
+        -----
+        The mean, scaling, and smoothing parameters are packed into
+        a vector.  Use `unpack` to access the component vectors.
+        """
+
+        mnpar, sdpar, smpar = self.unpack(params)
+        pm, pv = len(mnpar), len(sdpar)
+
+        # Residuals
+        resid = self.endog - np.dot(self.exog, mnpar)
+
+        # Standard deviations
+        sd = np.exp(np.dot(self.exog_scale, sdpar))
+
+        # Smoothness
+        sm = np.exp(np.dot(self.exog_smooth, smpar))
+
+        # Get the log-likelihood
+        score = np.zeros(len(mnpar) + len(sdpar) + len(smpar))
+        for _, ix in self._groups_ix.items():
+
+            # Get the covariance matrix for this person.
+            cm = self.cov.get_cov(self.time[ix], sd[ix], sm[ix])
+            cmi = np.linalg.inv(cm)
+
+            jacv, jacs = self.cov.jac(self.time[ix], sd[ix], sm[ix])
+
+            # The derivatives for the mean parameters.
+            dcr = np.linalg.solve(cm, resid[ix])
+            score[0:pm] += np.dot(self.exog[ix, :].T, dcr)
+
+            # The derivatives for the standard deviation parameters.
+            rx = np.outer(resid[ix], resid[ix])
```

Comment:
My guess is that repeated indexing here is costly.
It would be better to get the indexed array once and keep it during the iteration in the loop
e.g. use `resid_i = resid[ix]` at the beginning of the loop
similarly with other arrays that don't change during the loop iteration, e.g. sd[ix], sm[ix] are used 4 times each

(I'm spot checking not reviewing line by line)